File: app/src/main/java/com/digicoffer/lauditor/Webservice/CommonApiHelper/UploadImageTask.java
package com.digicoffer.lauditor.Webservice.CommonApiHelper;

import android.content.Context;
import android.os.AsyncTask;
import android.webkit.MimeTypeMap;

import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener;
import com.digicoffer.lauditor.Webservice.HttpResultDo;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants;

import org.apache.http.conn.ConnectTimeoutException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Iterator;

public class UploadImageTask extends AsyncTask<String, String, HttpResultDo> {

    private File uploadFile;
    private boolean sslFlag;
    private WebServiceHelper.RestMethodType restMethodType;
    private String baseURL;
    private AsyncTaskCompleteListener callback;
    private Context activity;
    private String requestType;

    public UploadImageTask(File uploadFile, boolean sslFlag, WebServiceHelper.RestMethodType restMethodType,
                           String baseURL, AsyncTaskCompleteListener callback, Context activity, String requestType) {
        this.uploadFile = uploadFile;
        this.sslFlag = sslFlag;
        this.restMethodType = restMethodType;
        this.baseURL = baseURL;
        this.callback = callback;
        this.activity = activity;
        this.requestType = requestType;
    }

    @Override
    protected HttpResultDo doInBackground(String... params) {
        HttpResultDo httpResult = new HttpResultDo();
        String boundary = "WebKitFormBoundary" + System.currentTimeMillis();
        String LINE_FEED = "\r\n";
        HttpURLConnection httpURLConnection = null;

        try {
            URL url;
            if (requestType.equals("Upload DocEditor File")) {
                url = new URL(baseURL);
            } else {
                url = new URL(Constants.base_URL + baseURL);
            }

            httpURLConnection = (HttpURLConnection) url.openConnection();
            httpURLConnection.setUseCaches(false);
            httpURLConnection.setDoOutput(true);
            httpURLConnection.setDoInput(true);
            httpURLConnection.setRequestMethod("POST");
            httpURLConnection.setRequestProperty("Cofferid", Constants.USER_ID);
            httpURLConnection.setRequestProperty("Authorization", "Bearer " + Constants.TOKEN);
            httpURLConnection.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + boundary);

            DataOutputStream requestStream = new DataOutputStream(httpURLConnection.getOutputStream());

            JSONObject json = new JSONObject(params[0] != null ? params[0] : "{}");
            Iterator<String> keys = json.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                String value = json.optString(key);

                requestStream.writeBytes("--" + boundary + LINE_FEED);
                requestStream.writeBytes("Content-Disposition: form-data; name=\"" + key + "\"" + LINE_FEED);
                requestStream.writeBytes("Content-Type: text/plain; charset=UTF-8" + LINE_FEED);
                requestStream.writeBytes(LINE_FEED);
                requestStream.writeBytes(value + LINE_FEED);
            }

            // Add content_type
            String contentType = getMimeType(uploadFile.getAbsolutePath());
            if (contentType == null) contentType = "application/octet-stream";

            requestStream.writeBytes("--" + boundary + LINE_FEED);
            requestStream.writeBytes("Content-Disposition: form-data; name=\"content_type\"" + LINE_FEED);
            requestStream.writeBytes("Content-Type: text/plain; charset=UTF-8" + LINE_FEED);
            requestStream.writeBytes(LINE_FEED);
            requestStream.writeBytes(contentType + LINE_FEED);

            // File part
            requestStream.writeBytes("--" + boundary + LINE_FEED);
            requestStream.writeBytes("Content-Disposition: form-data; name=\"file\"; filename=\"" + uploadFile.getName() + "\"" + LINE_FEED);
            requestStream.writeBytes("Content-Type: " + contentType + LINE_FEED);
            requestStream.writeBytes("Content-Transfer-Encoding: binary" + LINE_FEED);
            requestStream.writeBytes(LINE_FEED);

            FileInputStream inputStream = new FileInputStream(uploadFile);
            byte[] buffer = new byte[4096];
            int bytesRead;
            while ((bytesRead = inputStream.read(buffer)) != -1) {
                requestStream.write(buffer, 0, bytesRead);
            }
            inputStream.close();
            requestStream.writeBytes(LINE_FEED);

            requestStream.writeBytes("--" + boundary + "--" + LINE_FEED);
            requestStream.flush();
            requestStream.close();

            int status = httpURLConnection.getResponseCode();
            BufferedReader reader;
            if (status >= 200 && status < 300) {
                httpResult.setResult(WebServiceHelper.ServiceCallStatus.Success);
                reader = new BufferedReader(new InputStreamReader(httpURLConnection.getInputStream()));
            } else {
                httpResult.setResult(WebServiceHelper.ServiceCallStatus.Failed);
                reader = new BufferedReader(new InputStreamReader(httpURLConnection.getErrorStream()));
            }

            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line).append("\n");
            }
            reader.close();
            httpResult.setResponseContent(response.toString());

        } catch (ConnectTimeoutException e) {
            httpResult.setResult(WebServiceHelper.ServiceCallStatus.Exception);
            httpResult.setResponseContent("Exception: Connection Timeout " + e.getMessage());
        } catch (IOException e) {
            httpResult.setResult(WebServiceHelper.ServiceCallStatus.Exception);
            httpResult.setResponseContent("IOException: " + e.getMessage());
            AndroidUtils.logMsg("UploadImageTask: IO Exception " + e.getMessage());
        } catch (Exception e) {
            httpResult.setResult(WebServiceHelper.ServiceCallStatus.Exception);
            httpResult.setResponseContent(e.getMessage() != null ? e.getMessage() : "Unknown error");
            AndroidUtils.logMsg("UploadImageTask: Exception " + e.getMessage());
        } finally {
            if (httpURLConnection != null) {
                httpURLConnection.disconnect();
            }
        }

        return httpResult;
    }

    @Override
    protected void onPostExecute(HttpResultDo httpResult) {
        httpResult.setRequestType(requestType);
        super.onPostExecute(httpResult);
        try {
            callback.onAsyncTaskComplete(httpResult);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private String getMimeType(String filePath) {
        String extension = "";
        int index = filePath.lastIndexOf(".");
        if (index > 0) {
            extension = filePath.substring(index + 1).toLowerCase();
        }

        // Manual strong mapping
        switch (extension) {
            case "pdf":
                return "application/pdf";

            case "doc":
                return "application/msword";

            case "docx":
                return "application/vnd.openxmlformats-officedocument.wordprocessingml.document";

            case "xls":
                return "application/vnd.ms-excel";

            case "xlsx":
                return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";

            case "ppt":
                return "application/vnd.ms-powerpoint";

            case "pptx":
                return "application/vnd.openxmlformats-officedocument.presentationml.presentation";

            case "jpg":
            case "jpeg":
                return "image/jpeg";

            case "png":
                return "image/png";

            case "txt":
                return "text/plain";

            case "csv":
                return "text/csv";

            case "zip":
                return "application/zip";
        }

        // Secondary fallback
        String mime = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
        return mime != null ? mime : "application/";
    }

}

File: app/src/main/java/com/digicoffer/lauditor/Webservice/CommonApiHelper/UploadImageTask.kt
package com.digicoffer.lauditor.Webservice.CommonApiHelper

import android.content.Context
import android.os.AsyncTask
import android.webkit.MimeTypeMap
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener
import com.digicoffer.lauditor.Webservice.HttpResultDo
import com.digicoffer.lauditor.Webservice.core.RetrofitClient
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.util.HashMap

class UploadImageTask(
    private val uploadFile: File,
    private val sslFlag: Boolean,
    private val restMethodType: WebServiceHelper.RestMethodType,
    private val baseURL: String,
    private val callback: AsyncTaskCompleteListener,
    private val activity: Context,
    private val requestType: String
) : AsyncTask<String, String, HttpResultDo>() {

    override fun doInBackground(vararg params: String?): HttpResultDo {
        val httpResult = HttpResultDo()
        try {
            val url = if (requestType == "Upload DocEditor File") {
                baseURL
            } else {
                Constants.base_URL + baseURL
            }

            val headers = HashMap<String, String>()
            headers["Cofferid"] = Constants.USER_ID ?: ""
            headers["Authorization"] = "Bearer ${Constants.TOKEN}"

            val builder = MultipartBody.Builder().setType(MultipartBody.FORM)

            val paramsJson = if (params.isNotEmpty() && params[0] != null) params[0] else "{}"
            ApiMonitorLogger.logSending(
                requestType = requestType,
                url = url,
                method = "POST (Multipart)",
                params = "File: ${uploadFile.name} | Params: $paramsJson"
            )

            val json = JSONObject(paramsJson)
            val keys = json.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val value = json.optString(key)
                builder.addFormDataPart(key, value)
            }

            var contentType = getMimeType(uploadFile.absolutePath)
            if (contentType == null) {
                contentType = "application/octet-stream"
            }

            builder.addFormDataPart("content_type", contentType)

            val fileMediaType = contentType.toMediaTypeOrNull()
            val fileBody = uploadFile.asRequestBody(fileMediaType)
            builder.addFormDataPart("file", uploadFile.name, fileBody)

            val requestBody = builder.build()
            val call = RetrofitClient.getApiService().executePost(url, headers, requestBody)
            val response = call.execute()

            val statusCode = response.code()
            httpResult.status_code = statusCode

            val responseData = if (response.isSuccessful) {
                httpResult.result = WebServiceHelper.ServiceCallStatus.Success
                response.body()?.string() ?: ""
            } else {
                httpResult.result = WebServiceHelper.ServiceCallStatus.Failed
                response.errorBody()?.string() ?: ""
            }
            httpResult.responseContent = responseData

        } catch (e: IOException) {
            httpResult.result = WebServiceHelper.ServiceCallStatus.Exception
            httpResult.responseContent = "IOException: ${e.message}"
            AndroidUtils.logMsg("UploadImageTask: IO Exception ${e.message}")
        } catch (e: Exception) {
            httpResult.result = WebServiceHelper.ServiceCallStatus.Exception
            httpResult.responseContent = e.message ?: "Unknown error"
            AndroidUtils.logMsg("UploadImageTask: Exception ${e.message}")
        }

        return httpResult
    }

    override fun onPostExecute(result: HttpResultDo) {
        result.requestType = requestType
        super.onPostExecute(result)
        try {
            val url = if (requestType == "Upload DocEditor File") {
                baseURL
            } else {
                Constants.base_URL + baseURL
            }
            ApiMonitorLogger.logResponse(
                requestType = requestType,
                url = url,
                params = "File: ${uploadFile.name}",
                httpResult = result
            )
            callback.onAsyncTaskComplete(result)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun getMimeType(filePath: String): String? {
        val extension = filePath.substringAfterLast('.', "").lowercase()
        return when (extension) {
            "pdf" -> "application/pdf"
            "doc" -> "application/msword"
            "docx" -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            "xls" -> "application/vnd.ms-excel"
            "xlsx" -> "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            "ppt" -> "application/vnd.ms-powerpoint"
            "pptx" -> "application/vnd.openxmlformats-officedocument.presentationml.presentation"
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "txt" -> "text/plain"
            "csv" -> "text/csv"
            "zip" -> "application/zip"
            else -> MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension) ?: "application/octet-stream"
        }
    }
}
